ir/instr.cpp

```diff
@@ -1640,8 +1640,44 @@ void Return::print(ostream &os) const {
   os << val->getName();
 }
 
+static void addUBForNoCaptureRet(State &s, const expr &bid,
+    const expr &npptr, const StateValue &svret, const Type &t) {
+  auto &[vret, npret] = svret;
+  if (t.isPtrType()) {
+    s.addUB((npptr && npret).implies(
+        Pointer(s.getMemory(), vret).get_bid() != bid));
+    return;
+  }
+
+  if (auto agg = t.getAsAggregateType()) {
+    for (unsigned i = 0, e = agg->numElementsConst(); i != e; ++i) {
+      addUBForNoCaptureRet(s, bid, npptr, agg->extract(svret, i),
+                            agg->getChild(i));
+    }
+  }
+}
+
 StateValue Return::toSMT(State &s) const {
+  // Encode nocapture semantics.
+  vector<expr> nonpoisons;
+  vector<expr> bids;
+  for (const auto &I: s.getFn().getInputs()) {
+    auto inp = dynamic_cast<const Input *>(&I);
+    if (!inp->getType().isPtrType() ||
+        !inp->hasAttribute(Input::NoCapture))
+      continue;
+
+    auto &[v1, np1] = s[I];
+    auto bid = Pointer(s.getMemory(), v1).get_bid();
+    nonpoisons.emplace_back(np1);
+    bids.emplace_back(bid);
+
+    addUBForNoCaptureRet(s, bid, np1, s[*val], val->getType());
```

Comment:
this is reevaluating the return value many times, including the return stmt below